qa/rpc-tests/addressindex.py

```diff
@@ -0,0 +1,389 @@
+        for i in (0, 2):    # address b listed in args first and last
+            assert_equal(mempool[i]['address'], b)
+            assert_equal(mempool[i]['satoshis'], 3*COIN)
+            assert_equal(mempool[i]['txid'], txid)
+        # a single address can be specified as a string (not json object)
+        assert_equal([mempool[1]], self.nodes[0].getaddressmempool(a))
+
+        txids_a.append(txid)
+        expected_deltas.append({
+            'height': 111,
+            'satoshis': (-4)*COIN,
+            'txid': txid,
+        })
+        self.sync_all() # ensure transaction is included in the next block
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        # the send to b tx is now in a mined block, no longer in the mempool
+        mempool = self.nodes[0].getaddressmempool({'addresses': [b, a]})
+        assert_equal(len(mempool), 0)
+
+        # Test DisconnectBlock() by invalidating the most recent mined block
+        tip = self.nodes[1].getchaintips()[0]
+        for i in range(3):
+            bal = self.nodes[i].getaddressbalance(a)
+            assert_equal(bal['received'], expected*COIN)
+            # the value 4 UTXO is no longer in our balance
+            assert_equal(bal['balance'], (expected-4)*COIN)
+
+            bal = self.nodes[i].getaddressbalance(b)
+            assert_equal(bal['received'], 3*COIN)
+            assert_equal(bal['balance'], 3*COIN)
+
+            assert_equal(self.nodes[i].getblockcount(), 111)
+            self.nodes[i].invalidateblock(tip['hash'])
+            assert_equal(self.nodes[i].getblockcount(), 110)
+
+            mempool = self.nodes[i].getaddressmempool({'addresses': [b, a]})
+            assert_equal(len(mempool), 2)
+
+            bal = self.nodes[i].getaddressbalance(a)
+            assert_equal(bal['received'], expected*COIN)
+            assert_equal(bal['balance'], expected*COIN)
+
+            bal = self.nodes[i].getaddressbalance(b)
+            assert_equal(bal['received'], 0)
+            assert_equal(bal['balance'], 0)
+
+        # now re-mine the a to b send
+        self.nodes[0].generate(1)
+        self.sync_all()
+        for i in range(3):
+            assert_equal(self.nodes[i].getblockcount(), 111)
+
+        mempool = self.nodes[0].getaddressmempool({'addresses': [b, a]})
+        assert_equal(len(mempool), 0)
+
+        bal = self.nodes[2].getaddressbalance(a)
+        assert_equal(bal['received'], expected*COIN)
+        # the value 4 UTXO is no longer in our balance
+        assert_equal(bal['balance'], (expected-4)*COIN)
+
+        # Ensure the change from that transaction appears
+        tx = self.nodes[0].getrawtransaction(txid, 1)
+        change_vout = filter(lambda v: v['valueZat'] != 3*COIN, tx['vout'])
+        change = change_vout[0]['scriptPubKey']['addresses'][0]
+        bal = self.nodes[2].getaddressbalance(change)
+        assert(bal['received'] > 0)
+        assert(bal['received'] < (4-3)*COIN)
```

Comment:
Maybe add a comment here about why we do not check the exact amount.

Reply:
@eirik0, this change value is not consistent across multiple runs of this test. Usually, the balance is 99999774 zatoshis, but about one of 5 times, it's 99999775. Is that expected? Is there some randomness in the transaction fee? But at least for now, I'll add a comment, as you suggest.